pkg/policy/l4.go

```diff
@@ -1084,13 +1111,280 @@ func addL4Filter(policyCtx PolicyContext,
+	ForEach(func(l4 *L4Filter) bool)
+	Equals(t *testing.T, bMap L4PolicyMap) bool
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Len() int
+}
+
+// NewL4PolicyMap creates an new L4PolicMap.
+func NewL4PolicyMap() L4PolicyMap {
+	return &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+}
+
+// InitL4PolicyMap creates an new L4PolicMap, with an initial
+// set of values. The initMap argument does not support ranges.
+func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
+	l4M := &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+	for k, v := range initMap {
+		portProtoSlice := strings.Split(k, "/")
+		if len(portProtoSlice) < 2 {
+			continue
+		}
+		l4M.Upsert(portProtoSlice[0], 0, portProtoSlice[1], v)
+	}
+	return l4M
+}
+
+type portProtoKey struct {
+	port, endPort uint16
+	proto         uint8
+}
+
+// l4PolicyMap is the implementation of L4PolicyMap
+type l4PolicyMap struct {
+	namedPortMap map[string]*L4Filter
+	// rangePortMap has to keep a map of L4Filters rather than
+	// a single L4Filter reference so that the l4PolicyMap does
+	// not merge L4Filter that are not the same port range, but
+	// share an overlapping range in the trie.
+	rangePortMap *bitlpm.UintTrie[uint32, map[portProtoKey]*L4Filter]
+	// rangeMapLen must be tracked separately from
+	// rangePortMap as L4Filters are split up when
+	// the port range is long enough. This does not
+	// represent additional entries only the vagaries
+	// of indexing by range.
+	rangeMapLen int
+}
+
+func parsePortProtocol(port, protocol string) (uint16, uint8) {
+	// These string values have been validated many times
+	// over at this point.
+	prt, _ := strconv.ParseUint(port, 10, 16)
+	proto, _ := u8proto.ParseProtocol(protocol)
+	return uint16(prt), uint8(proto)
+}
+
+func makePolicyMapKey(port, mask uint16, proto uint8) uint32 {
+	return (uint32(proto) << 16) | uint32(port&mask)
+}
+
+// Upsert L4Filter adds an L4Filter indexed by protocol/port-endPort.
+func (l4M *l4PolicyMap) Upsert(port string, endPort uint16, protocol string, l4 *L4Filter) {
+	if iana.IsSvcName(port) {
+		l4M.namedPortMap[port+"/"+protocol] = l4
+		return
+	}
+
+	portU, protoU := parsePortProtocol(port, protocol)
+	ppK := portProtoKey{
+		port:    portU,
+		endPort: endPort,
+		proto:   protoU,
+	}
+	var upsertHappened bool
+	for _, mp := range PortRangeToMaskedPorts(portU, endPort) {
+		k := makePolicyMapKey(mp.port, mp.mask, protoU)
+		prefix := 32 - uint(bits.TrailingZeros16(mp.mask))
+		portProtoMap, ok := l4M.rangePortMap.ExactLookup(prefix, k)
+		if !ok {
+			portProtoMap = make(map[portProtoKey]*L4Filter)
+			l4M.rangePortMap.Upsert(prefix, k, portProtoMap)
+		}
+		if !upsertHappened {
+			if _, ok := portProtoMap[ppK]; !ok {
+				l4M.rangeMapLen += 1
+				upsertHappened = true
+			}
+		}
+		portProtoMap[ppK] = l4
```

Comment:
What happens if the user specifies two overlapping port ranges? It seems like this currently will just overwrite the initial upsert in a non-deterministic way.

Reply:
^ If this is only used with KNP, then the L4Filter is probably equivalent if two overlapping port ranges are injected. Then I guess if a CNP with L7 is added for a specific port, then that will be held in the other map, not in this port ranges map. So long as we properly encode the map entry for the specific port, that aspect should work. So perhaps the only awkward case would be if we enabled port ranges in CNP and allowed two overlapping port ranges with L7 fields under the L4 stanza...?

Reply:
What if the range is the same, but the L7 parameters are different? Will they be merged as you might expect?

Reply:
>What if the range is the same, but the L7 parameters are different? Will they be merged as you might expect?

Yes. In the `createFilter` method an exact range match is required to merge filters. We always know that unique ranges and ports are being upserted.

Reply:
Ah right yes, handled in the layers above. :+1: 

```
INCOMING CALLS
- f Upsert github.com/cilium/cilium/pkg/policy • l4.go
  - f addL4Filter github.com/cilium/cilium/pkg/policy • l4.go
    - f mergeIngressPortProto github.com/cilium/cilium/pkg/policy • rule.go
      - f mergeIngress github.com/cilium/cilium/pkg/policy • rule.go
        + f resolveIngressPolicy github.com/cilium/cilium/pkg/policy • rule.go
```

Reply:
I also looked into another corner case, if the start port is the same but the endport is different. That appears to be covered by indexing this map with a key of {port,endPort,proto}. I also see that `TestL4PolicyMapPortRangeOverlaps()` validates the map management aspects here with overlapping ranges. :+1: 

I didn't see unit tests for the outer layer that includes merging ranges with L7 fields. Maybe that's an area for subsequent improvement to reduce the risk of regression specifically related to those features. There are probably other testing aspects we should focus on first though.